File: services/rag-service/src/citation_builder.py
from __future__ import annotations

from typing import Any


def build_citations(chunks: list[dict[str, Any]]) -> list[dict[str, str]]:
    seen: set[str] = set()
    citations: list[dict[str, str]] = []
    for chunk in chunks:
        key = chunk["doc_id"]
        if key in seen:
            continue
        seen.add(key)
        citations.append(
            {
                "doc_id": chunk["doc_id"],
                "title": chunk["title"],
                "source_uri": chunk["source_uri"],
            }
        )
    return citations

File: services/rag-service/ts-src/citationBuilder.ts
export interface CitationChunk {
  readonly doc_id: string;
  readonly title: string;
  readonly source_uri: string;
  readonly chunk_id?: string;
}

export interface Citation {
  readonly doc_id: string;
  readonly title: string;
  readonly source_uri: string;
}

export function buildCitations(chunks: readonly CitationChunk[]): Citation[] {
  const seen = new Set<string>();
  const citations: Citation[] = [];
  for (const chunk of chunks) {
    const key = chunk.doc_id;
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    citations.push({
      doc_id: chunk.doc_id,
      title: chunk.title,
      source_uri: chunk.source_uri,
    });
  }
  return citations;
}
